gym_roboy/envs/roboy_env.py

```diff
@@ -0,0 +1,132 @@
+import numpy as np
+import gym
+from gym import spaces
+from . import ROSProxy
+from .robots import RobotState, RoboyRobot
+
+
+def _l2_distance(joint_angle1, joint_angle2):
+    subtract = np.subtract(joint_angle1, joint_angle2)
+    subtract[np.isnan(subtract)] = 0  # np.inf - np.inf returns np.nan, but should be 0
+    return np.linalg.norm(subtract, ord=2)
+
+
+class RoboyEnv(gym.GoalEnv):
+
+    def __init__(self, ros_proxy: ROSProxy, seed: int = None,
+                 joint_vel_penalty: bool = False,
+                 is_tendon_vel_dependent_on_distance: bool = True,
+                 is_agent_getting_bonus_for_reaching_goal: bool = True):
+        self.seed(seed)
+        self._ros_proxy = ros_proxy
+        self._joint_vel_penalty = joint_vel_penalty
+        self._is_tendon_vel_dependent_on_distance = is_tendon_vel_dependent_on_distance
+        self._is_agent_getting_bonus_for_reaching_goal = is_agent_getting_bonus_for_reaching_goal
+        self._robot = robot = ros_proxy.robot
+
+        self._GOAL_JOINT_VEL = robot.new_zero_state().joint_angle
+        self._MAX_DISTANCE_JOINT_ANGLE = _l2_distance(robot.get_joint_angles_space().low, robot.get_joint_angles_space().high)
+        self._MAX_DISTANCE_JOINT_VELS = _l2_distance(robot.get_joint_vels_space().low, robot.get_joint_vels_space().high)
+        self._PENALTY_FOR_TOUCHING_BOUNDARY = 1
+        self._BONUS_FOR_REACHING_GOAL = 10
+
+        self.reward_range = self._create_reward_range(robot=robot)
+        self.action_space = spaces.Box(low=-1, high=1, shape=robot.get_action_space().shape, dtype="float32")
+        self.observation_space = spaces.Box(
+            low=np.concatenate((robot.get_joint_angles_space().low, robot.get_joint_vels_space().low, robot.get_joint_angles_space().low)),
+            high=np.concatenate((robot.get_joint_angles_space().high, robot.get_joint_vels_space().high, robot.get_joint_angles_space().high)),
+        )
+        self._set_new_goal()
+
+    def _create_reward_range(self, robot: RoboyRobot) -> Tuple[float, float]:
+        some_state = robot.new_state(
+            joint_angle=robot.get_joint_angles_space().high,
+            joint_vel=robot.get_joint_vels_space().high, is_feasible=True)
+        corresponding_worst_state = robot.new_state(
+            joint_angle=robot.get_joint_angles_space().low,
+            joint_vel=robot.get_joint_vels_space().low, is_feasible=False)
+        max_reward = self.compute_reward(current_state=some_state, goal_state=some_state)
+        min_reward = self.compute_reward(current_state=corresponding_worst_state, goal_state=some_state)
+        return min_reward, max_reward
+
+    def step(self, action):
+        assert self.action_space.contains(action)
+
+        if self._is_tendon_vel_dependent_on_distance:
+            distance_to_target = _l2_distance(self._last_state.joint_angle, self._goal_state.joint_angle)
+            scale_factor = np.power(distance_to_target / self._MAX_DISTANCE_JOINT_ANGLE, 1/4)
```

Comment:
can you explain this?

Reply:
we want the agent to move slower as it approaches the goal. we measure that with the fraction `scale_factor = distance_to_target / MAX_DISTANCE_JOINT_ANGLE` which goes from 0 to 1.
Since our boundaries on python are only a fraction of the real boundaries of the robot, our `scale_factor` was constantly in the region from 0 to 0.1. So our robot was moving very slowly from the beginning. We decided to change the `scale_factor` to be x^1/4 of the old one. Effectively this increases the speed at the region from 0 to 0.1 while still reducing the scale_factor to 0 when the robot reaches the goal.